### PR TITLE
ADM-960 [frontend]: finish the logic code and test

### PR DESCRIPTION
**Pull request by zhou-yinyuan**
## Summary

Charts module adaptation and loading.

## Before

When we only select `cycle time` metrics

image

image

## After

image

image

## Note

_Null_

### CODE REVIEW

File: frontend/e2e/fixtures/import-file/partial-metrics-show-chart.ts [hunk 1]
```diff
@@ -0,0 +1,100 @@
+export const partialMetricsShowChart = {
+  projectName: 'Heartbeat Metrics',
```
Comment (guzhongren) [inline frontend/e2e/fixtures/import-file/partial-metrics-show-chart.ts:2]: update project name to the test case name, will be more efficient

File: frontend/e2e/fixtures/input-files/partial-metrics-show-chart.template.json [hunk 2]
```diff
@@ -0,0 +1,96 @@
+{
+  "projectName": "Heartbeat Metrics",
```
Comment (guzhongren) [inline frontend/e2e/fixtures/input-files/partial-metrics-show-chart.template.json:2]: same issue

File: frontend/src/containers/ReportStep/DoraMetricsChart/index.tsx [hunk 3]
```diff
@@ -162,53 +164,134 @@ export const DoraMetricsChart = ({ data, dateRanges }: DoraMetricsChartProps) =>
...
   });
 
+  const leadTimeForChangesValuelist = mappedData
+    .flatMap((value) => value.leadTimeForChangesList)
+    .filter((value) => value?.name === AVERAGE)
+    .map((value) => value?.valuesList);
+
+  const isLeadTimeForChangesFinished =
+    leadTimeForChangesValuelist.length === dateRanges.length &&
+    leadTimeForChangesValuelist.every((value) => value?.every((it) => it.value != ''));
```
Comment (guzhongren) [inline frontend/src/containers/ReportStep/DoraMetricsChart/index.tsx:174]: use `EMPTY_STRING` to replace `''`

File: frontend/src/containers/ReportStep/DoraMetricsChart/index.tsx [hunk 4]
```diff
@@ -162,53 +164,134 @@ export const DoraMetricsChart = ({ data, dateRanges }: DoraMetricsChartProps) =>
...
+
+  const deploymentFrequencyValueList = mappedData
+    .flatMap((value) => value.deploymentFrequencyList)
+    .filter((value) => value?.name == AVERAGE)
+    .map((value) => value?.valueList);
+  const isDeploymentFrequencyFinished =
+    deploymentFrequencyValueList.length == dateRanges.length &&
+    deploymentFrequencyValueList.every((value) => value?.every((it) => it.value != ''));
```
Comment (guzhongren) [inline frontend/src/containers/ReportStep/DoraMetricsChart/index.tsx:182]: duplicated code

File: frontend/src/containers/ReportStep/index.tsx [hunk 5]
```diff
@@ -85,6 +85,11 @@ export interface DateRangeRequestResult {
   reportData: ReportResponseDTO | undefined;
 }
 
+export const CHART_LOADING = {
+  text: '',
+  color: 'rgba(255, 255, 255, 0.8)',
```
Comment (guzhongren) [inline frontend/src/containers/ReportStep/index.tsx:90]: have u passed the pipeline check?

File: ops/check.sh [hunk 6]
```diff
@@ -103,6 +103,7 @@ rgba_check() {
     --exclude-dir='node_modules' \
     --exclude-dir='coverage' \
     --exclude='*.html' \
+    --exclude='index.tsx' \
```
Comment (guzhongren) [inline ops/check.sh:106]: revert it! 
`No rgba` is our convention.

**Review by guzhongren — CHANGES_REQUESTED** [review]

File: ops/check.sh (hunk 6 above)
Comment (zhou-yinyuan) [reply ops/check.sh]: done

File: frontend/src/containers/ReportStep/index.tsx (hunk 5 above)
Comment (zhou-yinyuan) [reply frontend/src/containers/ReportStep/index.tsx]: I move it to theme.ts

File: frontend/e2e/fixtures/import-file/partial-metrics-show-chart.ts (hunk 1 above)
Comment (zhou-yinyuan) [reply frontend/e2e/fixtures/import-file/partial-metrics-show-chart.ts:2]: done

File: frontend/e2e/fixtures/input-files/partial-metrics-show-chart.template.json (hunk 2 above)
Comment (zhou-yinyuan) [reply frontend/e2e/fixtures/input-files/partial-metrics-show-chart.template.json:2]: done

File: frontend/src/containers/ReportStep/DoraMetricsChart/index.tsx (hunk 3 above)
Comment (zhou-yinyuan) [reply frontend/src/containers/ReportStep/DoraMetricsChart/index.tsx]: done

File: frontend/src/containers/ReportStep/DoraMetricsChart/index.tsx (hunk 4 above)
Comment (zhou-yinyuan) [reply frontend/src/containers/ReportStep/DoraMetricsChart/index.tsx]: refactored

**Review by Mandy-Tang — APPROVED** [review]

File: frontend/src/containers/ReportStep/DoraMetricsChart/index.tsx [hunk 7]
```diff
@@ -148,7 +157,39 @@ function extractedMeanTimeToRecoveryDataData(allDateRanges: string[], mappedData
...
+function isDoraMetricsChartFinish({
+  dateRangeLength,
+  mappedData,
+  type,
+}: {
+  dateRangeLength: number;
+  mappedData: (
+    | ReportResponse
+    | {
+        deploymentFrequencyList: { id: number; name: string; valueList: { value: string }[] }[];
```
Comment (guzhongren) [inline frontend/src/containers/ReportStep/DoraMetricsChart/index.tsx]: Think: replace those `key` with the enum `DORAMetricsChartType`

**Review by guzhongren — APPROVED** [review]

File: frontend/src/containers/ReportStep/DoraMetricsChart/index.tsx (hunk 7 above)
Comment (zhou-yinyuan) [reply frontend/src/containers/ReportStep/DoraMetricsChart/index.tsx]: done

File: frontend/__tests__/containers/ReportStep/ReportStep.test.tsx [hunk 8]
```diff
@@ -691,6 +691,63 @@ describe('Report Step', () => {
...
+      dispatchAction: jest.fn(),
+      dispose: jest.fn(),
+      showLoading: jest.fn(),
+      hideLoading: jest.fn(),
+    };
+    beforeEach(() => {
+      jest.spyOn(echarts, 'init').mockImplementation(
+        () =>
+          //eslint-disable-next-line @typescript-eslint/no-explicit-any
+          chart as any,
```
[the remark this answers is not in the document]
Comment (zhou-yinyuan) [reply frontend/__tests__/containers/ReportStep/ReportStep.test.tsx]: but in previous test code, the any is used in mock test

**Pull request merged**